### PR TITLE
[JENKINS-24937] Add a JDK configuration on promotion status

### PR DESCRIPTION
**Pull request by ehartmann**
This improvement permit the definition of JDK for the promotion.
When using a MatrixProject without this, it's impossible to execute Java process such as maven task.
This fix https://issues.jenkins-ci.org/browse/JENKINS-24937

### CODE REVIEW

File: src/main/java/hudson/plugins/promoted_builds/PromotionProcess.java [hunk 1]
```diff
@@ -69,12 +69,25 @@
...
+     * @see Jenkins#getJDK(String)
+     */
+    private String jdk;
+
     /**
      * The label that promotion process can be run on.
      */
     public String assignedLabel;
-    
+
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/promoted_builds/PromotionProcess.java]: please avoid space-only changes

File: src/main/java/hudson/plugins/promoted_builds/PromotionProcess.java [hunk 2]
```diff
@@ -203,7 +216,11 @@ public String getAssignedLabelString() {
     }
 
     @Override public JDK getJDK() {
-        return getOwner().getJDK();
+        if (jdk == null) {
+          return getOwner().getJDK();
+        } else {
+          return Jenkins.getInstance().getJDK(jdk);
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/promoted_builds/PromotionProcess.java]: This code may return null. There should be a fall-back to the default JDK

**Comment by oleg-nenashev** [comment]
JDK is a special case in Jenkins due to historical reasons. I would not vote for directly expanding it, but it may be OK if somebody find this feature useful

**Comment by oleg-nenashev** [comment]
Needs a formatting fix before the merge

File: src/main/resources/hudson/plugins/promoted_builds/PromotionProcess/process-config.jelly [hunk 3]
```diff
@@ -37,6 +37,17 @@
         </select>
       </f:entry>
 
+      <j:set var="jdks" value="${app.JDKs}" />
```
Comment (oleg-nenashev) [inline src/main/resources/hudson/plugins/promoted_builds/PromotionProcess/process-config.jelly]: Should be moved to advanced options IMHO

**Comment by oleg-nenashev** [comment]
@ehartmann 
Any plans to finalize the PR?

**Comment by ehartmann** [comment]
@oleg-nenashev, Sorry I was off. I'll try to do make the changes asap.

**Comment by oleg-nenashev** [comment]
@ehartmann 
Hi, thanks a lot. Currently I'm integrating the changes for promoted-builds-2.22. My plan is to release the beta version on this week

**Comment by ehartmann** [comment]
@oleg-nenashev : I've made the following modifications : 
* Readd the spaces missing,
* Move JDK configuration to advanced,
* Test and fix an issue where the JDK is updated on both the promotion process and the main project.

About potential null value for JDK, I'm not sure how to handle this case. The list of JDKs are retrieved from installed ones, so if there is a null value, it seems that default JDK is used. How do you want to handle this case ?

**Comment by oleg-nenashev** [comment]
> About potential null value for JDK, I'm not sure how to handle this case. The list of JDKs are retrieved from installed ones, so if there is a null value, it seems that default JDK is used. How do you want to handle this case ?

Yes, <code>AbstractProject:hetJDK()</code> may return null according to Javadoc. In such case there is no need to handle this case in your code.

File: src/main/resources/hudson/plugins/promoted_builds/PromotionProcess/process-config.jelly [hunk 4]
```diff
@@ -37,6 +37,19 @@
...
+      <f:advanced>
+        <f:entry title="JDK"
+                 description="${%JDK to be used for this project}">
+          <select class="setting-input validated" name="jdkForPromotion" checkUrl="'${rootURL}/defaultJDKCheck?value='+this.value">
+            <f:option selected="${(inst == null) || (inst.name=='UseOwner')}" value="UseOwner">${%UseSameAsJob}</f:option>
```
Comment (oleg-nenashev) [inline src/main/resources/hudson/plugins/promoted_builds/PromotionProcess/process-config.jelly]: Probably makes sense to add a help file to describe the behavior and the magic <code>UseOwner</code> string. A common way would be to just use empty/null string to specify this default behavior.

**Pull request closed without merging**